libraries/botbuilder-core/src/cloudAdapterBase.ts

```diff
@@ -0,0 +1,356 @@
+    ClaimsIdentity,
+    ConnectorClient,
+    ConnectorFactory,
+    UserTokenClient,
+} from 'botframework-connector';
+
+import {
+    Activity,
+    ActivityEx,
+    ActivityTypes,
+    Channels,
+    ConversationReference,
+    DeliveryModes,
+    InvokeResponse,
+    ResourceResponse,
+    StatusCodes,
+} from 'botframework-schema';
+
+export abstract class CloudAdapterBase extends BotAdapter {
+    public readonly ConnectorFactoryKey = Symbol('ConnectorFactory');
+    public readonly UserTokenClientKey = Symbol('UserTokenClient');
+
+    /**
+     * Create a new [CloudAdapterBase](xref:botbuilder.CloudAdapterBase) instance.
+     *
+     * @param botFrameworkAuthentication A [BotFrameworkAuthentication](xref:botframework-connector.BotFrameworkAuthentication) used for validating and creating tokens.
+     */
+    constructor(protected readonly botFrameworkAuthentication: BotFrameworkAuthentication) {
+        super();
+
+        if (!botFrameworkAuthentication) {
+            throw new TypeError('`botFrameworkAuthentication` parameter required');
+        }
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public sendActivities(context: TurnContext, activities: Partial<Activity>[]): Promise<ResourceResponse[]> {
+        if (!context) {
+            throw new TypeError('`context` parameter required');
+        }
+
+        if (!activities) {
+            throw new TypeError('`activities` parameter required');
+        }
+
+        if (!activities.length) {
+            throw new Error('Expecting one or more activities, but the array was empty.');
+        }
+
+        return Promise.all(
+            activities.map(async (activity) => {
+                delete activity.id;
+
+                if (activity.type === 'delay') {
+                    await delay(typeof activity.value === 'number' ? activity.value : 1000);
+                } else if (activity.type === ActivityTypes.InvokeResponse) {
+                    context.turnState.set(INVOKE_RESPONSE_KEY, activity);
+                } else if (activity.type === ActivityTypes.Trace && activity.channelId !== Channels.Emulator) {
+                    // no-op
+                } else {
+                    const connectorClient = context.turnState.get<ConnectorClient>(this.ConnectorClientKey);
+                    if (!connectorClient) {
+                        throw new Error('Unable to extract ConnectorClient from turn context.');
+                    }
+
+                    if (activity.replyToId) {
+                        return connectorClient.conversations.replyToActivity(
+                            activity.conversation.id,
+                            activity.replyToId,
+                            activity
+                        );
+                    } else {
+                        return connectorClient.conversations.sendToConversation(activity.conversation.id, activity);
+                    }
+                }
+
+                return { id: activity.id ?? '' };
```

Comment:
I checked the [.NET implementation](https://github.com/microsoft/botbuilder-dotnet/blob/475c222551521dbae758f5d0d0a9f3fbb39175fb/libraries/Microsoft.Bot.Builder/CloudAdapterBase.cs#L107) to make sure, but isn't this _always_ equivalent to `{ id: '' }` because of the `delete activity.id;` call?

In .NET and here, the `activity.id` was deleted at the start of the loop.

If I'm reading this right, when the outgoing activity's type is Delay, an InvokeResponse, or a Trace with a non-Emulator channel, the `ResourceResponse` object always has an empty string for `ResourceResponse.id`.

Reply:
I thought so too if I'm honest 🤔